packages/nimble-components/src/table/specs/action-menu-hld.md

```diff
@@ -0,0 +1,113 @@
+```
+
+### Providing a menu
+
+Because only one menu can be open on the page at a time, the client can provide a single menu that will be associated with the correct menu button when it is open. The menu is provided by the client by slotting a menu in the `action-menu` slot as shown below:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-text field-name="firstName" show-menu>First name</nimble-table-column-text>
+    <nimble-table-column-text field-name="lastName">Last name</nimble-table-column-text>
+
+    <nimble-menu slot="action-menu">
+        <nimble-menu-item>My first action</nimble-menu-item>
+        <nimble-menu-item>My second action</nimble-menu-item>
+        <nimble-menu-item>My last action</nimble-menu-item>
+    </nimble-menu>
+</nimble-table>
+```
+
+If an application requires different menu items for different rows, the client is responsible for ensuring that the items in the menu are correct for the row(s) that the menu is associated. This can be done by handling the `action-menu-opening` event on the table and updating the menu items as appropriate. The `action-menu-opening` event will include the following in its details:
+
+-   `rowIds` - string array - The IDs of the rows that the menu is associated with.
+-   `columnId` - string, possibly undefined - The ID of the column that the menu is associated with. A column ID is optional on a column definition. If the menu is associated with a column without an ID, `columnId` will be `undefined` in the event details.
+
+### Implementation
+
+#### Slot Forwarding
+
+In order for the menu slotted within the table to be slotted within a cell's `nimble-menu-button`, the menu needs to be "forwarded" from the table to a row and then to a cell. Every row will have a slot for an menu, but only the row with the open menu will have something slotted within it. Similarly, every cell in a column with a menu will have a slot for a menu, but only the cell with an open menu will have something slotted within it. Events will be used to know when the row and cell the menu is associated with needs to change, and the templates will dynamically rename their slots to ensure the menu is slotted in the correct elements.
+
+For example, the table's template will look something like this:
+
+```HTML
+<template>
+    ...
+    <nimble-table-row>
+        <slot slot="row-action-menu" name="${(x, c) => ((c.parent.openActionMenuRowId === x.id) ? 'action-menu' : 'unused-action-menu')}"></slot>
+    </nimble-table-row>
+    ...
+</template>
+```
+
+#### Updates to `nimble-menu-button`
+
+To implement the design described above, a few small changes need to be made to the existing `nimble-menu-button`:
+
+-   Add an `opening` event that gets fired immediately before the menu is opened. This new event will allow the table to slot the menu into the correct row and cell prior to the menu actually opening. This is important to ensure that the menu items can be focused correctly upon the menu opening.
```

Comment:
> This is important to ensure that the menu items can be focused correctly upon the menu opening.

I'm a bit concerned that this might be difficult to achieve generically. I don't think that all the frameworks will be able to guarantee that the response to the event (instantiating some DOM / running some templates) will be synchronous with the event firing and guarantee that it happens before the event resolves.

For example, something like blazor server will receive the event, send the event to the server,resolve the event,  and then respond based on the server action. It'll be quick, in 10s of ms, but definitely after the event synchronously resolves. Similar concerns even for angular where a template can listen for events but observed behavior will be evaluated only during change detection (which I'm not confident when that runs, may be a microtask or on rAF, etc).

Do we have prototypes with Blazor / Angular that tries out this system?

(without yet searching around / inventing an api on the spot) It may be that we need a custom event, ie `MenuOpeninEvent` and we add a method like `waitFor(promise)` where the listener for the opening event can call that method and pass a promise that when resolved by the user will notify the menu that it actually can open. When called the method sets some state on the event (similar to how preventDefault(), etc work) that we can check for and avoid opening the menu until that promise resolves.

Reply:
Another thought, maybe those that can do it synchronously do so and those that can't are required to `focus()` the element themselves when they actually render. An issue there is we may end-up with menu content visually popping in compared to the promise approach only showing when ready

Reply:
Maybe keep it straightforward in nimble itself. Make the event cancellable, the simple usage in vanilla js will just do dom setup in the callback no issue.

Frameworks that cannot respond synchronously (Blazor and maybe Angular) can cancel the event, finish their work, and then manually set open on the menu-button.

Maybe we have a service in nimble angular / helper in blazor to manage that state easier (listen for open events, ignore events / disable the app thinks menu is "ready", then set open on the menu button).

Looking at the menu-button it looks like changing the open property value will trigger an event which outside any of this is likely a bug. We should only fire state change events in response to user interactions, not programmatic property changes.

Reply:
"manually set open on the menu-button" conceptually, not sure about the right api for that in practice. Might need to be something on the table like `openActionMenu(index)`

Reply:
I had similar questions when trying to imagine what a typical implementation of this event handler might look like.

In a naive approach an application might programmatically add/remove menu items with DOM APIs. But that would be pretty ugly since it'd be writing DOM code in JS/TS and not in the template language of the framework. But to Milan's point, at least it would be synchronous.

A cleaner approach might be to have the event handler just update some state on the component that hosts the table. If the component's template binds to that state to control which set of menu-items is shown, hopefully the framework would update the UI before the menu actually appears. But I'm guessing that wouldn't work if change detection isn't synchronous?

Reply:
A thought that came to mind that may even be less fiddly for clients, what if we let multiple action menu definitions that can be defined ahead of time?

```html
<nimble-table>
   <nimble-column-text action-menu="my-menu-1">
   <nimble-column-text action-menu="my-menu-1">
   <nimble-column-text action-menu="my-menu-2">

   <menu slot="my-menu-1"></menu>
   <menu slot="my-menu-2"></menu>
</nimble-table>
```

I don't think 90% of usage requires that synchronous completely menu generation per row and column. My guess / proposal is that all rows for a particular column will likely have the same action menu to show, so let it be defined in the dom ahead of time.

If you absolutely need logic that is row value specific in the rendering of the menu then you have to respond to the event synchronously. But most usages will render their menu stuff ahead of time, when the user clicks we synchronously show that menu, let the framework respond to the events and clicks of the menu on it's time (framework will get the event, cache off the current row and column and use that to respond to button clicks, etc).

Also nice because you can have it be much more declarative in static html, just need js to respond to button clicks, etc instead of dynamically rendering content.

Reply:
I tested the action menu in Blazor and Angular, and I didn't run into any timing issues responding to the `action-menu-opening` event and updating the set of items shown in the menu. Specifically, I dynamically generated a set of options in HTML using `for` in Blazor and `ngFor` in Angular where the set was updated in the `action-menu-opening` callback. I was able to use the up & down arrows to open the menu and have the correct item focused. I also didn't see any flickering of items being added/removed from the menu.

Given this assessment, I'm going to leave the events as they are proposed in the spec and not try to do anything fancy with asynchronous event handling. I do, however, like the suggestion of having a different action menu for each column, so I'll update the spec with that API change.